ingestion_server/ingestion_server/api.py

```diff
@@ -128,40 +134,53 @@ def on_get(self, req, resp):
 class TaskStatus(BaseTaskResource):
     def on_get(self, req, resp, task_id):
         """Check the status of a single task."""
-        task = self.tracker.id_task[task_id]
-        active = task.is_alive()
+        try:
+            task = self.tracker.id_task[task_id]
+        except KeyError:
+            resp.status = falcon.HTTP_404
+            resp.media = {"message": f"No task found with id {task_id}"}
+            return
 
         percent_completed = self.tracker.id_progress[task_id].value
+        active_workers = bool(self.tracker.id_active_workers[task_id].value)
+        active = task.is_alive() or active_workers
+
         resp.media = {
             "active": active,
             "percent_completed": percent_completed,
             "error": percent_completed < 100 and not active,
         }
 
 
-class WorkerFinishedResource:
+class WorkerFinishedResource(BaseTaskResource):
     """
     For notifying ingestion server that an indexing worker has finished its
     task.
     """
 
-    @staticmethod
-    def on_post(req, _):
-        target_index = worker_finished(str(req.remote_addr))
-        if target_index:
+    def on_post(self, req, _):
+        task_data = worker_finished(str(req.remote_addr), req.media["error"])
+        task_id = task_data["task_id"]
+        # Update progress
+        self.tracker.id_progress[task_id] = task_data["percent_completed"]
+
+        if task_data["percent_completed"] == 100:
             logging.info(
                 "All indexer workers finished! Attempting to promote index "
-                f"{target_index}"
+                f"{task_data['target_index']}"
```

Comment:
Nit: since `task_data["target_index"]` is used a few times in this block, it might be worth setting it to a variable